### PR TITLE
Fixing Linq Performance Issue for Navigation binding cache

### PR DESCRIPTION
**Pull request by Sreejithpin**
### Issues
Fixing Linq Performance Issue for Navigation binding cache

*This pull request fixes issue #xxx.*

### Description

*Briefly describe the changes of this pull request.*

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: src/Microsoft.OData.Core/UriParser/Aggregation/ApplyBinder.cs [hunk 1]
```diff
@@ -125,7 +125,12 @@ private static IEnumerable<AggregateTokenBase> MergeEntitySetAggregates(IEnumera
                 }
             }
 
-            return mergedTokens.Concat(entitySetTokens.Values).ToList();
+            foreach(AggregateTokenBase tokenBase in entitySetTokens.Values)
+            {
+                mergedTokens.Add(tokenBase);
+            }
```
Comment (marabooy) [inline src/Microsoft.OData.Core/UriParser/Aggregation/ApplyBinder.cs:131]:
```suggestion   
                mergedTokens.AddRange(entitySetTokens.Values);
```

File: src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs [hunk 2]
```diff
@@ -155,7 +155,12 @@ internal static bool TryBindAsOperation(PathSegmentToken pathToken, IEdmModel mo
                     NonSystemToken nonSystemToken = pathToken as NonSystemToken;
                     if (nonSystemToken != null && nonSystemToken.NamedValues != null)
                     {
-                        parameterNames = nonSystemToken.NamedValues.Select(s => s.Name).ToList();
+                        parameterNames = new List<string>();
+
+                        foreach(NamedValue namedValue in nonSystemToken.NamedValues)
```
Comment (marabooy) [inline src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs:160]: The linq query on the right could be faster in most cases as the list may need to resize during addition.

File: src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs [hunk 3]
```diff
@@ -130,11 +130,20 @@ public IEnumerable<IEdmNavigationPropertyBinding> FindNavigationPropertyBindings
...
                 if (!navigationPropertyBindingCache.TryGetValue(navigationProperty, out navigationBindingList))
                 {
-                    navigationBindingList = this.NavigationPropertyBindings.Where(targetMapping => targetMapping.NavigationProperty == navigationProperty);
+                    navigationBindingList = new List<IEdmNavigationPropertyBinding>();
+
+                    foreach (IEdmNavigationPropertyBinding targetMapping in this.NavigationPropertyBindings)
+                    {
+                        if (targetMapping.NavigationProperty == navigationProperty)
+                        {
+                            navigationBindingList.Add(targetMapping);
```
Comment (marabooy) [inline src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs:143]: Add range should be faster in most cases as it uses a faster array copy method for the items.

File: src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs (hunk 2 above)
Comment (Sreejithpin) [reply src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs:160]: But it does 2 iteration right. 1 for select and another for T0List

Comment (KenitoInc) [reply src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs:160]:
```suggestion
                        foreach (NamedValue namedValue in nonSystemToken.NamedValues)

```

**Review by KenitoInc — CHANGES_REQUESTED** [review]

File: src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs (hunk 3 above)
Comment (Sreejithpin) [reply src/Microsoft.OData.Edm/Csdl/Semantics/CsdlSemanticsNavigationSource.cs:143]: You mean add range with a where to check for navigationproperty? that would be slower right?

File: src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs (hunk 2 above)
Comment (habbes) [reply src/Microsoft.OData.Core/UriParser/Binders/SelectPathSegmentTokenBinder.cs:160]: @Sreejithpin  does it really do 2 iterations? I think `.Select()` is lazily evaluated when `ToList()` is called, and probably leads to just one iteration? I'd be curious to see the benchmarks for this particular loop compared to the code being replaced.

**Comment by habbes** [comment]
@Sreejithpin maybe you could share a screenshot of the benchmark results for these changes?

**Comment by Sreejithpin** [comment]
Closing as it was already done in https://github.com/OData/odata.net/pull/2123

**Pull request closed without merging**